### ISSUE TITLE
Make tests build cacheable

### ISSUE THREAD

**Issue by martinbonnin** [JetBrains/intellij-platform-gradle-plugin#1430]
Tests are not build cacheable. Running them twice in a row re-runs them. To reproduce:

```
git clone https://github.com/apollographql/apollo-kotlin
cd apollo-kotlin
git checkout 4824ae3bf53025a0b2b6290edb609647834f435c
# run tests a first time
./gradlew :intellij-plugin:test 
# delete outputs
rm -rf intellij-plugin/build
# re-run tests
./gradlew :intellij-plugin:test 
# tests are executed again
```

The second test invocation should reuse the cached results but instead the tests are re-run. 

I'm suspecting it's because some of the input/output directories are shared between tasks or something like this? On some occasions, I have seen logs like this:

```
Caching disabled for task ':intellij-plugin:test' because:
  Gradle does not know how file 'build/idea-sandbox/system-test/log/idea.log' was created (output property 'System directory'). Task output caching requires exclusive access to output paths to guarantee correctness (i.e. multiple tasks are not allowed to produce output in the same location).
```

Other times, it was `classpath.index` being missing in one of the invocations. 

Tests currently take ~1min so it's not the end of the world but would be nice to get those minutes back.

**Comment by YannCebron** [comment]
there could be changes in
- testData files used in tests (e.g., highlighting)
- any external systems (e.g., LSP)

that are impossible to track

**Comment by martinbonnin** [comment]
> testData files used in tests

We can track that manually:

```kotlin
tasks.withType<Test> {
  inputs.dir("testData").withPropertyName("testData").withPathSensitivity(PathSensitivity.RELATIVE)
}
```

> any external systems (e.g., LSP)

Not overly familiar with LSP and/or other things. Is that something IntelliJ downloads at startup? If yes that could be an issue but maybe something that could be pinned/disabled using configuration?

**Comment by YannCebron** [comment]
testData directories could be located anywhere, or even generated at runtime

LSP server is just a sample for any external system that tests may depend on. Or e.g. some external linter that is called via `ExternalAnnotator` and many other things.

**Comment by martinbonnin** [comment]
My superficial understanding of the IntelliJ IDE is that it's mostly an executable .jar and therefore self contained/predictable and ultimately cacheable?

I understand IntelliJ is highly customizable and as such could take inputs from "the internet" that are impossible to track but these customizations would not be important for our tests so feels like having a way to run the IDE without them for reproducable tests would be a win?

**Comment by YannCebron** [comment]
This is about what the actual tests of the plugin project do. We basically can't predict it or set "useful" defaults for caching.

Reference for test data: https://plugins.jetbrains.com/docs/intellij/test-project-and-testdata-directories.html

**Comment by YannCebron** [comment]
That's correct, but a breaking change for any plugin relying on the previous behaviour (or having non-default testData location(s)).

**Comment by hsz** [comment]
We re-checked this on the current 2.x plugin (IntelliJ Platform Plugin Template).

**The 1.x cache-disable is gone.** In 1.x the sandbox `system`/`log` directories were declared as task outputs, which produced:

> Caching disabled for task ':…:test' because: Gradle does not know how file 'build/idea-sandbox/system-test/log/idea.log' was created (output property 'System directory'). Task output caching requires exclusive access to output paths…

In 2.x those directories are `@Internal` and are only passed to the test JVM as `-Didea.system.path` / `-Didea.log.path`. The sandbox itself also moved out of `build/` to `.intellijPlatform/sandbox/<project>/<product-version>/` (test logs live in `log-test/idea.log`).

**Verified:** with the build cache on, `./gradlew test --build-cache`, then `rm -rf build/` (keeping `.intellijPlatform/`), then `./gradlew test --build-cache` again → `:test FROM-CACHE`. No “Caching disabled” / “System directory” / “exclusive access to output paths” on `:test`.

The standard `test` task is still Gradle’s own `Test` type. Custom `testIde` tasks remain `@UntrackedTask` on purpose (`Should always run`) — that does not apply to built-in `test`.

If your tests read extra files (e.g. `src/test/testData`) and you *don’t* get cache hits, declare those inputs yourself:

```kotlin
tasks.test {
    inputs.dir(layout.projectDirectory.dir("src/test/testData"))
        .withPropertyName("testData")
        .withPathSensitivity(PathSensitivity.RELATIVE)
        .optional()
}
```

Closing as addressed by the 2.x sandbox redesign. If `test` still fails to cache on 2.x with inputs declared, please open a new issue with a current repro.
